[waspls] diagnostics for external imports and goto definition

craigmc08 opened the pull request:
- Adds diagnostics for extimports trying to import JS code that does not exist
- Adds goto definition for extimports
- Refactors LSP to support a little bit of multithreading

infomiho commented on waspc/packages/ts-inspect/README.md:
```diff
@@ -0,0 +1,26 @@
+  { "filenames": ["./src/exports.ts"] },
+  {
+    "tsconfig": "~/dev/wasp-todoapp/src/client/tsconfig.json",
+    "filenames": "~/dev/wasp-todoapp/src/client/MainPage.tsx"
```
@craigmc08 I'm a bit confused about the API here.

There is a key called `filenames` but it only has a single string? Is this a mistake in writing the README?

craigmc08 replied:
That's a mistake in the README, should be an array. Fixed!

infomiho commented on waspc/packages/ts-inspect/src/index.ts:
```diff
@@ -0,0 +1,36 @@
+  if (!Array.isArray(input)) {
+    throw new Error('Expected JSON array from stdin');
+  }
+  const requests: ExportsRequest[] = input;
```
I would recommend defining a schema with Zod and parsing the input so you can be sure about the format 100% https://zod.dev/?id=basic-usage

infomiho commented on waspc/packages/ts-inspect/src/exports.ts:
```diff
@@ -0,0 +1,77 @@
+
+  // Loop through each given file and try to get its exports.
+  for (let filename of filenames) {
+    const source = program.getSourceFile(filename);
```
I would maybe extract the body of the loop into a helper function `getExportsForFile` and maybe do some `throw`ing and `catch`ing in the loop to output the errors and continue the loop instead of using `continue`

infomiho commented on waspc/packages/ts-inspect/test/exportTests/complex.ts:
```diff
@@ -0,0 +1,15 @@
+export default function isEvenOrOdd(n: number): boolean {
```
What about exporting constants, do you differentiate those from functions?

```ts
export const name = "Craig";
```

craigmc08 replied:
Exporting a constant is counted as a named export, it is not differentiated from functions. I added a new test case to document/check this behavior.

infomiho commented on waspc/packages/ts-inspect/src/exports.ts:
```diff
@@ -0,0 +1,93 @@
+    } catch (err) {
+      console.error(err);
+      exportsMap[filename] = [];
+      continue;
```
@craigmc08 I think `continue` doesn't do anything here, the loop continues irregardless

craigmc08 replied:
Deleted `continue`

infomiho commented on waspc/waspls/test/Wasp/LSP/DebouncerTest.hs:
```diff
@@ -0,0 +1,54 @@
+    mvar <- newEmptyMVar
+
+    debounce debouncer 1000 () (void $ tryPutMVar mvar ())
+    threadDelay 20000
```
Does this mean that our unit tests have a minimum time of running of 20 seconds? 😄

craigmc08 replied:
Thankfully not! `threadDelay` takes a delay in microseconds, so the longest test is only 40ms.

infomiho commented:
Tested it locally, it feels really cool to be able to jump to definition from the Wasp file. I'll let @Martinsos do the review of the Haskell part.

- One thing that feels a bit weird: the popup with `w` or `g` above the definition (I'm guessing that's the first letter), can we avoid that?
- Another thing that feels weird is the fact I can hover over multiple things and click, but they take me to the same place. Maybe reduce the number of places for which we offer the goto definition feature if possible?

Uploading Screen Recording 2023-06-22 at 16.19.32.mov…

craigmc08 commented:
The single letter popups does seem like the first letter. I think that's because I respond with a single-character selection range in the TS file. I'll see if I can get the typescript compiler to give me more information on the definition.

The hover and click showing different underline areas -> I think that's because I don't specify what area the link covers, so VSCode assumes it is whatever word you're hovering over. Fixing this should be relatively quick.

Martinsos reviewed (approved):
@craigmc08  amazing job here, this is looking great!
I think all together it looks very sane and makes a lot of sense.

One big thing is rethinking the `Package` name, another is just ensuring that we keep LSP as understandable as possible -> it obviously got more complicated with these changes, which is to some degree unavoidable beacuse we have global state and threads, but if we can thighten it up even a bit more, that is a plus.

I would recommend following -> that we merge this PR, so we can get it into the Wasp 0.11, but that you also continue working on it immediately after it, and we then also merge the improvements relatively soon once they are ready.

I am not sure how to best handle this technically -> do we merge this PR, and you keep looking at it and we comment on these old comments on closed PR while you open a new one with changes?
Or do we create a branch from main called `main-for-1268`, have this PR point to it (while also merging it into `main`), and then continue working on it? Not sure, whatever you think is the best!

craigmc08 commented:
Ok I will merge this then.

Here's my plan for handling the merge while keeping the review comments: I'll merge this branch normally and create a new PR addressing feedback here + more code clean up. As I address the feedback with code changes, I'll quote comment your feedback on the modified code and link to those new comments in this PR. That way we have a way of checking that I responded to everything. Does that plan work for you?

Martinsos commented:
> Ok I will merge this then.
> 
> Here's my plan for handling the merge while keeping the review comments: I'll merge this branch normally and create a new PR addressing feedback here + more code clean up. As I address the feedback with code changes, I'll quote comment your feedback on the modified code and link to those new comments in this PR. That way we have a way of checking that I responded to everything. Does that plan work for you?

Sure, go for it!

craigmc08 commented on waspc/waspls/src/Wasp/LSP/Debouncer.hs:
```diff
@@ -0,0 +1,44 @@
+-- fires.
+debounce :: (MonadUnliftIO m, MonadIO m, Eq k, Hashable k) => Debouncer k -> Int -> k -> m () -> m ()
+debounce (Debouncer running) waitMicros event fire = do
+  fireIO <- toIO fire
+  a <- liftIO $
+    async $ do
+      threadDelay waitMicros
+      fireIO
+      atomically $ STM.delete event running
+  prev <- liftIO $
+    atomically $ do
+      prev <- STM.lookup event running
+      STM.insert a event running
+      return prev
+  liftIO $ traverse_ cancel prev
```
🚀 https://github.com/wasp-lang/wasp/pull/1292/files#diff-d2463dff307c7af8f8dc13d40e1c035b53d940ed41dd80232f655dfa333bf8a3R52-R73

craigmc08 commented on waspc/waspls/src/Wasp/LSP/Diagnostic.hs:
```diff
@@ -1,38 +1,76 @@
+data WaspDiagnostic
+  = ParseDiagnostic !CPE.ParseError
+  | AnalyzerDiagonstic !W.AnalyzeError
+  | MissingImportDiagnostic !SourceSpan !MissingImportReason !(SP.Path' SP.Abs SP.File')
```
Good point! Renamed to `MissingExtImport`: https://github.com/wasp-lang/wasp/pull/1292/files#diff-0c7b958f2ee2116a36caf6e6a5509bd1d5befb21d309a669a22ed6a117b3e814R25

craigmc08 reviewed (commented):
Responded to your feedback @Martinsos and gave links to the changes in the new draft PR: https://github.com/wasp-lang/wasp/pull/1292.

There are some comment threads in here that will take some discussion to resolve. Let's keep the old discussions in here, but start any new discussions ones in the new PR.

No rush on another round of review, I still have more of your feedback to work on that will take longer, just got most of the low-hanging fruit for now. I'll ping you when it's ready 😄

Martinsos commented on waspc/packages/ts-inspect/src/index.ts:
```diff
@@ -0,0 +1,28 @@
+import { ExportRequests, getExportsOfFiles } from "./exports.js";
+
+async function readStdin(): Promise<string> {
```
Whatever you like better! If what we have works for sure and we are not sure about fs.readFile(0), we can stick with what we have. Maaaaaybe you could rename it to `readAllFromStdin` -> but you will have a better feeling for this than me.

Martinsos commented on waspc/src/Wasp/Package.hs:
```diff
@@ -0,0 +1,91 @@
+--
+-- If the package does not have its dependencies installed yet (i.e. after they
+-- just installed a Wasp version), we install the dependencies.
+getPackageProc :: Package -> [String] -> IO P.CreateProcess
```
Ok all makes sense!
The thing is, naming is really bad on the `System.Process` side. `CreateProcess` -> pretty bad name, would be better if it was `ProcessCreator` or smth like that.
I forgot that that function is called `proc`, so seeing `getPackageProc` didn't help me much, I wasn't sure what `proc` their means. So I wouldn't stick to how they are calling it, instead would try to come up with a better name.

Ideas:
- `preparePackageForCalling` -> I like that it hides info about what it returns -> one can check that in the type. But, it also sounds like it does something every time, while it doesn't, it will do npm install and stuff only if needed, it doesn't do heavy prepartion every time.
- `ensurePackageIsPreparedForCalling` -> A bit better, `ensure` for me indicates it will ensure some conditions, which might need it will just confirm them, or actually produce them. But it sounds like it doesn't return anything. Which is not terrible, but still.
- `createPackageProcess` -> Plays on the idea that it returns `CreateProcess`. More haskellish, because if you give it two args, it inded becomes create process for package, but I think it also hides quite a bit so I don't like that a lot. `CReateProcess` name is giving us trouble here.
- `getPackageProcessCreator` -> Hides details about preparing the package, but makes it clear what it returns. Could be ok, caller doesn't really need to know how happens in the background. Maybe a better name would be `obtainPackageProcessCreator` -> `obtain` indicates it might need some heavy lifting, unlike `get` which usually means there is not much going on.

Looking at all this, maybe `preparePackageForCalling` or `obtainPackageProcessCreator` are the best in my opinion.

craigmc08 replied:
I renamed it to `getPackageProcessOptions`. I think there isn't much of a difference between get and obtain here, so I just chose the shorter one. And "process options" seems, to me, like a much better name than `CreateProcess`, because it literally is the options/settings that `createProcess` uses when it spawns a process.

Martinsos commented on waspc/src/Wasp/TypeScript.hs:
```diff
@@ -0,0 +1,114 @@
+
+-- | A list of files associated with an optional tsconfig file that is run
+-- through the TypeScript compiler as a group.
+data TsExportRequest = TsExportRequest {filenames :: ![FilePath], tsconfig :: !(Maybe FilePath)}
```
bump

Martinsos commented on waspc/src/Wasp/TypeScript.hs:
```diff
@@ -0,0 +1,114 @@
+      (_ :: Value) -> fail "invalid type for TsExport"
+
+-- | Map from TypeScript files to the list of exports found in that file.
+newtype TsExportResponse = TsExportResponse (M.HashMap FilePath [TsExport])
```
bump

Martinsos commented on waspc/waspls/src/Wasp/LSP/Debouncer.hs:
```diff
@@ -0,0 +1,44 @@
+-- | @debounce debouncer waitMicros event action@ waits @waitMicros@ microseconds
+-- and then runs @action@.
+--
+-- If 'debounce' is called again with the same @event@, only the newer call
+-- fires.
```
Maybe you could mention then that old event is cancelled, and that timer is reset.

Martinsos commented on waspc/waspls/src/Wasp/LSP/Debouncer.hs:
```diff
@@ -0,0 +1,44 @@
+-- has its own debounce timer. Construct a debouncer with 'newDebouncerIO'.
+--
+-- See 'debounce' for how to use it.
+newtype Debouncer k = Debouncer (STM.Map k (Async ()))
```
Makes sense!

Martinsos commented on waspc/waspls/src/Wasp/LSP/ExtImport.hs:
bump

Martinsos commented on waspc/waspls/src/Wasp/LSP/GotoDefinition.hs:
```diff
@@ -0,0 +1,79 @@
+-- | @link linkRange location@ creates a @LSP.LocationLink@ to the same place as
+-- @location@ and sets the origin selection range (the range that is highlighted
+-- in the editor in the original file) to @linkRange@.
+link :: LSP.Range -> LSP.Location -> LSP.LocationLink
```
bump

craigmc08 replied:
It's a more convenient constructor for `LSP.LocationLink`. So I could rename it to `locationLink` but I'm not sure if that's more descriptive, since the type tells you that it returns a `LSP.LocationLink` anyways.

Martinsos commented on waspc/waspls/src/Wasp/LSP/ServerM.hs:
```diff
@@ -1,85 +1,115 @@
-liftLSP m = ServerM $ lift $ lift m
+-- | Most LSP handlers should use this instead of 'ServerM', as there are only
+-- limited places where modifying the state is needed.
+type HandlerM = RLspM ServerState
```
Yeah, maybe renaming then!

craigmc08 commented on waspc/waspls/src/Wasp/LSP/Handlers.hs:
Moved the code that was cluttering this file to new files:
- Dynamic capability registering is in its own place now (DynamicHandlers).
- Doing analysis/diagnostics on the wasp file also has its own file now.
There's still a lot of comments on the dynamic capability stuff, but it just needs a lot of context from the LSP spec to explain what it's doing.

craigmc08 commented on waspc/src/Wasp/Package.hs:
```diff
@@ -0,0 +1,91 @@
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Wasp.Package
```
Renamed to `NodePackageFFI`. I didn't change how the variables are named inside the module, it all just refers to "package" since the module name gives the context.

> Are we ok with tying this just to Node?

I think tying it to node is better for now, especially since the code relies heavily on them being node programs. If we end up wanting to run code from other languages we can just rename the module again. #

craigmc08 commented on waspc/packages/README.md:
I think since we went with `Wasp.NodePackageFFI`, it's ok to just keep this folder named `packages`. `js/npm/ts-packages` isn't much more descriptive.
